File: macOS/Sources/AudioCaptureKit/Models/CaptureError.swift
import Foundation

/// Errors that can occur during audio capture operations.
public enum CaptureError: Error, Sendable, Equatable {
    /// The user denied the required system permission.
    case permissionDenied

    /// The requested audio device is not available.
    case deviceNotAvailable

    /// The capture session could not be configured with the given parameters.
    case configurationFailed(String)

    /// Audio encoding failed during processing.
    case encodingFailed(String)

    /// Encryption of audio data failed.
    case encryptionFailed(String)

    /// An error occurred writing to or reading from storage.
    case storageError(String)

    /// The operation timed out.
    case timeout

    /// An unknown or unexpected error occurred.
    case unknown(String)
}

extension CaptureError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .permissionDenied:
            return "Permission denied. Please grant the required audio recording permission."
        case .deviceNotAvailable:
            return "The requested audio device is not available."
        case .configurationFailed(let reason):
            return "Configuration failed: \(reason)"
        case .encodingFailed(let reason):
            return "Encoding failed: \(reason)"
        case .encryptionFailed(let reason):
            return "Encryption failed: \(reason)"
        case .storageError(let reason):
            return "Storage error: \(reason)"
        case .timeout:
            return "The operation timed out."
        case .unknown(let reason):
            return "Unknown error: \(reason)"
        }
    }
}

File: csharp/AudioCapture/Models/CaptureError.cs
namespace AudioCapture.Models;

/// <summary>
/// The kind of capture error, matching Swift CaptureError variants.
/// </summary>
public enum CaptureErrorKind
{
    PermissionDenied,
    DeviceNotAvailable,
    ConfigurationFailed,
    EncodingFailed,
    EncryptionFailed,
    StorageError,
    Timeout,
    Unknown,
}

/// <summary>
/// Exception type for audio capture errors. Mirrors Swift CaptureError.
/// </summary>
public class CaptureException : Exception
{
    public CaptureErrorKind ErrorKind { get; }

    public CaptureException(CaptureErrorKind kind, string message, Exception? inner = null)
        : base(message, inner)
    {
        ErrorKind = kind;
    }

    public static CaptureException PermissionDenied(string message = "Microphone permission denied") =>
        new(CaptureErrorKind.PermissionDenied, message);

    public static CaptureException DeviceNotAvailable(string message = "Audio device not available") =>
        new(CaptureErrorKind.DeviceNotAvailable, message);

    public static CaptureException ConfigurationFailed(string message) =>
        new(CaptureErrorKind.ConfigurationFailed, message);

    public static CaptureException EncodingFailed(string message) =>
        new(CaptureErrorKind.EncodingFailed, message);

    public static CaptureException EncryptionFailed(string message) =>
        new(CaptureErrorKind.EncryptionFailed, message);

    public static CaptureException StorageError(string message) =>
        new(CaptureErrorKind.StorageError, message);

    public static CaptureException TimeoutError(string message = "Capture timed out") =>
        new(CaptureErrorKind.Timeout, message);

    public static CaptureException Unknown(string message) =>
        new(CaptureErrorKind.Unknown, message);
}
